third_party/txt/src/txt/paragraph.cc

```diff
@@ -318,20 +319,68 @@ void Paragraph::Layout(double width, bool force) {
 
     size_t layout_start = run.start;
     // Layout until the end of the run or too many lines.
-    while (layout_start < run.end && lines_ < paragraph_style_.max_lines) {
+    while (layout_start < run.end && lines_ < max_lines) {
       const size_t next_break = (break_index > breaks_count - 1)
                                     ? std::numeric_limits<size_t>::max()
                                     : breaks[break_index];
       const size_t layout_end = std::min(run.end, next_break);
 
       bool bidiFlags = paragraph_style_.rtl;
+      std::shared_ptr<minikin::FontCollection> minikin_font_collection =
+          font_collection_->GetMinikinFontCollectionForFamily(
+              run.style.font_family);
+
+      uint16_t* text_ptr = text.data() + layout_start;
+      size_t text_count = layout_end - layout_start;
+      std::vector<uint16_t> ellipsized_text;
+
+      // Apply ellipsizing if the run was not completely laid out and this
+      // is the last line (or lines are unlimited).
+      const std::u16string& ellipsis = paragraph_style_.ellipsis;
+      if (ellipsis.length() && !isinf(width_) && run.end != layout_end &&
```

Comment:
The complexity of this method is getting a bit high. In a later patch, we should see how to simplify this.

Reply:
Yes - it would be great to simplify Paragraph::Layout.  There might be some easy wins here, e.g. moving postprocess_line into a separate method